slay/definitions.py

```diff
@@ -0,0 +1,171 @@
+import abc
+from types import GenericAlias
+from typing import Any, ClassVar, Generic, Mapping, Optional, Type, TypeVar
+
+import pydantic
+
+UserConfigT = TypeVar("UserConfigT", bound=Optional[pydantic.BaseModel])
+
+BASTEN_APY_SECRET_NAME = "baseten_api_key"
+ENDPOINT_NAME = "run"  # Referring to processor method name exposed as endpoint.
+# Below arg names must correspond to `definitions.ABCProcessor`.
+CONTEXT_ARG_NAME = "context"  # Referring to processors `__init__` signature.
+SELF_ARG_NAME = "self"
+
+
+class APIDefinitonError(TypeError):
+    """Raised when user-defined processors do not adhere to API constraints."""
+
+
+class MissingDependencyError(TypeError):
+    """Raised when a needed resource could not be found or is not defined."""
+
+
+class UsageError(Exception):
+    """Raised when components are not used the expected way at runtime."""
+
+
+class Image(pydantic.BaseModel):
+    # TODO: this is a placeholder/dummy object.
+
+    def pip_requirements_txt(self, *args, **kwargs) -> "Image":
+        return self
+
+    def pip_install(self, *args, **kwargs) -> "Image":
+        return self
+
+    def cuda(self, *args, **kwargs) -> "Image":
+        return self
+
+
+class Resources(pydantic.BaseModel):
+    # TODO: this is a placeholder/dummy object.
+
+    def cpu(self, *args, **kwargs) -> "Resources":
+        return self
+
+    def gpu(self, *args, **kwargs) -> "Resources":
+        return self
+
+
+class Config(pydantic.BaseModel, Generic[UserConfigT]):
```

Comment:
Is the goal for this ultimately be 1-1 with [TrussConfig](https://github.com/basetenlabs/truss/blob/main/truss/truss_config.py#L323)? I worry about having some drift here, so maybe as a fast-follow we could consolidate these types

Reply:
1:1 is not the goal ;)

I wanted to decouple those types, to allow for a fresh thinking and structuring of what we want to make configurable (and how) for processors. Since (atm) we use truss as a "vehicle" to run processors this will be converted to a truss config and there is a lot of resemblance. But I think it's good to add this abstraction layer here to give use more flexibility on how we want this to look eventually.

It touches a bit on the question whether we (long-term) want workflows & processors to completely replace the current style of truss models or whether we want to create a dedicated server type for processors...